### PR TITLE
Split out immutable GadtConstraint

### PR DESCRIPTION
**Pull request by dwijnand**
This change eagerly allocates a GadtConstraint whenever its state is changed, rather than simply changing the reference(s) in, what's now called, `GadtConstraintHandling`, such as `constraint`.  Previous to a recent change, in order to restore the GADT constraints, a fresh copy was eagerly created in a number of cases, just so it can be used to restore, for the cases in which it must be restored.  The recent change exposed the underlying pieces of the GADT constraint and allowed those to be restored as components.  Now we still do that, but as a packaged up GadtConstraint that we created once, eagerly.  That also makes sure that invariants in components are upheld, like the mappings.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/GadtConstraint.scala [hunk 1]
```diff
@@ -2,34 +2,145 @@ package dotty.tools
...
+end GadtConstraint
+
+object GadtConstraintHandling:
+  def apply(gadt: GadtConstraint): GadtConstraintHandling = new ProperGadtConstraintHandling(gadt)
+
+sealed trait GadtConstraintHandling(private var myGadt: GadtConstraint) {
+  this: ConstraintHandling =>
+
+  def gadt: GadtConstraint              = myGadt
+  private def gadt_=(g: GadtConstraint) = myGadt = g
```
Comment (Linyxus) [inline compiler/src/dotty/tools/dotc/core/GadtConstraint.scala:143]: Maybe we should not make `GadtConstraint` a private variable in `GadtConstraintHandling`, but declare two methods for retrieving and updating GADT constraints in this trait, and require the class that extends this trait to provide this two methods, just like what we do in `ConstraintHandling` ([here](https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala#L36-L37)). Then `Context.fresh` would do the job of `GadtConstraint.fresh` as long as we put the `GadtConstraint` in the context.

**Review by Linyxus — CHANGES_REQUESTED** [review]
Otherwise LGTM! 👍

**Comment by dwijnand** [comment]
> Shall we have a `GadtConstraint` in the `Context` and extends `GadtConstraintHandling` wherever we would like to manipulate GADT constraints?
>
> Maybe we should not make `GadtConstraint` a private variable in `GadtConstraintHandling`, but declare two methods for retrieving and updating GADT constraints in this trait, and require the class that extends this trait to provide this two methods, just like what we do in `ConstraintHandling` ([here](https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala?rgh-link-date=2023-01-24T00%3A48%3A58Z#L36-L37)). Then `GadtConstraintHandling` will be a stateless mixin for GADT constraint manipulation, and `Context.fresh` would do the job of `GadtConstraint.fresh` as long as we put the `GadtConstraint` in the context.

So that doesn't work, because we want code that's running in an inner context to have changes apply to an outer context. 
 It's similar to TyperState.   To that end I renamed it to GadtState, which is shorter and more palatable.

**Review by Linyxus — APPROVED** [review]
Thanks for the explanation. LGTM!

**Pull request merged**